### ISSUE TITLE
Make string interpolation and `builtins.toString` behave the same

### ISSUE THREAD

**Issue by fricklerhandwerk** [NixOS/nix#7327]
**Is your feature request related to a problem? Please describe.**

`toString exp` and `"${exp}"` behave differently.

> `toString` *e*
> 
> Convert the expression *e* to a string. *e* can be:
> 
>   - A string (in which case the string is returned unmodified).
>   - A path (e.g., `toString /foo/bar` yields `"/foo/bar"`.
>   - A set containing `{ __toString = self: ...; }` or `{ outPath = ...; }`.
>   - An integer.
>   - A list, in which case the string representations of its elements
>     are joined with spaces.
>   - A Boolean (`false` yields `""`, `true` yields `"1"`).
>   - `null`, which yields the empty string.

Only the first three work with string interpolation, but not e.g. `"${1}"` or `"${true}"`.

**Describe the solution you'd like**

The following should work:

```nix
"${1}"
```

    "1"
    

```nix
"${[ 1 /foo/bar true ]}"
```

    "1 /nix/store/<hash>-bar true"

```nix
"${true}"
```

    "true"

```nix
"${false}"
```

    ""
    
```nix
"${null}"
```

    ""
    

**Describe alternatives you've considered**
Make `toString` work on fewer things, to make the features converge consistency, but that would break existing code or require a deprecation strategy.

**Additional context**
It's not clear why these differences even exist, except that string interpolation on paths serves a different purpose than for the other data types.

**Comment by tejing1** [comment]
At least make integers work... the rest are of more debatable value, but integers working in antiquotes is just basic functionality in any decent language at this point.

**Comment by sternenseemann** [comment]
> Only the first three work with string interpolation, but not e.g. `"${1}"` or `"${true}"`.

String interpolation has different behavior from `toString` when it comes to paths. Making both more similar will only increase the confusion around it.

Additionally, implicit type conversions are usually a source of pitfalls, so it is actually good that string interpolation is limited to the absolute essentials (i.e. Nix would be painful to use if paths and derivations needed explicit conversion).

**Comment by tejing1** [comment]
I would argue that the current behavior of `toString` on paths is broken design in the first place. It should be difficult, perhaps even impossible, for nix code to extract eval-time file paths as data.

Path types vs. strings encode a fundamental difference in how the information is being treated (eval time vs run time, essentially), and having basic language constructs completely obliterate that difference just confuses the issue for newcomers, inviting them to make an entire class of mistakes, and annoys the experienced.

**Comment by lalishansh** [comment]
Hi @sternenseemann, @delroth, @tejing1, @fricklerhandwerk, @mbbx6spp ji,
Previously i was using `pathType = types.coercedTo types.path (p: builtins.toString p) types.str;` in order to get absolute path to file in my nix-dotfiles, i've recently trying to upgrade so found out, this old way (now feels like a hack) dosen't work now and  i'm unable to make symlinks to orignal files, do you know how i can get absolute paths from files (types.path) ?

basically i am trying to do
```
users.users.ilal = {
# BLAH-BLAH-BLAH
    path.".config/hypr/hyprland.conf".link = ./hyprland.conf;
};
```
and ` "${./hyprland.conf}" ` now returns `/nix/store/xxxxxxxxxxxxxxxx-hyprland.conf`
